db/impl/consul/lock.go

```diff
@@ -0,0 +1,240 @@
+func (c *Client) lock(obj db.Lock, ttl time.Duration, refresh bool) (*lock, bool, error) {
+	// consul has a minimum ttl of 10s. Adjust.
+	if ttl < 10*time.Second {
+		ttl = 10 * time.Second
+	}
+
+	path, err := obj.Path()
+	if err != nil {
+		return nil, false, errors.LockFailed.Combine(err)
+	}
+
+	tmplock, ok := c.getLock(path)
+	if ok {
+		if reflect.DeepEqual(tmplock.obj, obj) {
+			return tmplock, true, nil
+		}
+
+		return nil, false, errors.LockFailed.Combine(errored.Errorf("Invalid lock attempted at %q -- already exists", path))
+	}
+
+	sessionID, err := c.getSession(obj, ttl, refresh)
+	if err != nil {
+		return nil, false, err
+	}
+
+	content, err := jsonio.Write(obj)
+	if err != nil {
+		return nil, false, err
+	}
+
+	mylock, err := c.client.LockOpts(&api.LockOptions{
+		LockTryOnce:      true,
+		Key:              c.qualified(path),
+		Value:            content,
+		SessionTTL:       ttl.String(),
+		MonitorRetryTime: ttl,
+		MonitorRetries:   -1,
+	})
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	stopChan := make(chan struct{})
+	go func() {
+		time.Sleep(100 * time.Millisecond)
+		close(stopChan)
+	}()
+
+	monitor, err := mylock.Lock(stopChan)
+	if err != nil {
+		return nil, false, err
+	}
+
+	done := make(chan struct{})
+
+	lockObj := &lock{
+		lock:        mylock,
+		sessionID:   sessionID,
+		doneChan:    done,
+		monitorChan: monitor,
+		obj:         obj,
+		ttl:         ttl,
+		path:        path,
+		refresh:     refresh,
+	}
+
+	return lockObj, false, nil
+}
+
+// doRefresh hangs and lets the consul client refresh its key at 1/4 TTL.
+func (c *Client) doRefresh(lockObj *lock) {
+	if err := c.client.Session().RenewPeriodic((lockObj.ttl / 4).String(), lockObj.sessionID, nil, lockObj.doneChan); err != nil {
+		logrus.Errorf("Error during periodic refresh of lock: %v", err)
+	}
+}
+
+// monitorLock monitors a lock (intended to be run in a goroutine). It blocks
+// until it receives one of several signals to terminate the lock refresh, or
+// delete the lock if the ttl has expired.
+func (c *Client) monitorLock(lockObj *lock) {
+	after := make(<-chan time.Time)
+
+	if !lockObj.refresh {
+		after = time.After(lockObj.ttl)
+	}
+
+	select {
+	case <-after:
+		lockObj.lock.Unlock()
+	case <-lockObj.monitorChan:
+	case <-lockObj.doneChan:
+		if err := c.Free(lockObj.obj, false); err != nil {
+			logrus.Error(err)
+		}
+	}
+
+	c.lockMutex.Lock()
+	lock, ok := c.locks[lockObj.path]
+	if ok && reflect.DeepEqual(lock.obj, lockObj.obj) {
+		delete(c.locks, lockObj.path)
+	}
+	c.lockMutex.Unlock()
+}
+
+// doAcquire does most of the work for the acquire* class of functions in the
+// consul client. See the other private functions in this file for more
+// information, but this is usually the top of them all.
+func (c *Client) doAcquire(obj db.Lock, ttl time.Duration, refresh bool) (chan struct{}, error) {
+	logrus.Debugf("Attempting to acquire %v", obj)
+
+	lockObj, running, err := c.lock(obj, ttl, refresh)
```

Comment:
This might again be rare but i think a race might exist here i.e. two routines attempting to lock same `obj` might both succeed in getting a new lock object of there own as the `c.locks` data structure get's updated much later below.